### PR TITLE
Assignment a

### PR DESCRIPTION
**Pull request by DevonScoles**
This PR contains tests for all create, read, update, and delete operations.

Tests to be skipped:
test_application_response_if_user_isnt_logged_in()
test_application_response_if_user_is_logged_in()
I would skip these because they are already tested via the auth folder inside tests which comes with every new Laravel project.
All other tests inside the NoteTest.php file I feel are necessary since they test the core functions of the application.

### CODE REVIEW

File: app/Http/Controllers/NoteController.php [hunk 1]
```diff
@@ -14,7 +14,7 @@ public function index()
     {
```
Comment (fw-aaron) [inline app/Http/Controllers/NoteController.php:14]: (This comment doesn't have to do with this line of code. I'm just creating a "conversation" that can be "resolved".)

I'm trying to run this. I ran the following:

```bash
sudo su;
apt update && apt install -y zip unzip git npm;
cd /app;
php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');";
php -r "if (hash_file('sha384', 'composer-setup.php') === 'dac665fdc30fdd8ec78b38b9800061b4150413ff2e3b6f88543c636f7cd84f6db9189d43a81e5503cda447da73c7e5b6') { echo 'Installer verified'; } else { echo 'Installer corrupt'; unlink('composer-setup.php'); } echo PHP_EOL;";
php composer-setup.php;
php -r "unlink('composer-setup.php');";
mv composer.phar /usr/bin/composer;
composer install;
sed 's/^APP_URL.*/APP_URL=http:\/\/localhost:8000/' .env.example > .env;
php artisan key:generate;
php artisan migrate;
npm i;
php artisan serve --host 0.0.0.0;
```
When I visit http://localhost:8000, I get

`Vite manifest not found at: /app/public/build/manifest.json`

Comment (fw-aaron) [reply app/Http/Controllers/NoteController.php:14]: Oh, I was able to get around that by running `npm run dev` as you instructed!

File: tests/Feature/MyTests/NoteTest.php [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            $userData[] = [
+                'name' => 'Test User #' . $i,
+                'email' => 'testemail' . $i . '@test.com',
+                'password' => bcrypt('pass123.'),
+            ];
+        }
+
+        User::factory()->createMany($userData);
+        $this->userOne = User::find(1); //assign userOne
+        $this->userTwo = User::find(2);//assign userTwo
```
Comment (fw-aaron) [inline tests/Feature/MyTests/NoteTest.php]: You're getting the two users by primary key, but you didn't assign a primary key above. In other words, you don't know that these are the users you created! (If a user was created before this function, you would be grabbing those earlier-created users.) Get them by name or e-mail address or something to make sure you're getting the records you expect.

File: tests/Feature/MyTests/NoteTest.php [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        $response->assertStatus(302);
+
+        $response = $this->get('/login');
+        $response->assertStatus(200);
+    }
+
+    public function test_application_response_if_user_is_logged_in()
+    {
+        $this->actingAs($this->userOne);
+        $response = $this->get('/note');
```
Comment (fw-aaron) [inline tests/Feature/MyTests/NoteTest.php:92]: Earlier (in the `isnt_logged_in` test case), `/` and `/login` are tested. Here, `/note` is tested. I think it would be more instructive if `/note` were tested in the `isnt_logged_in` testcase to show the difference.

File: tests/Feature/MyTests/NoteTest.php [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        $response = $this->get('/note/10');
+        $response->assertStatus(200);
+        $response->assertSee('this is note #10 for user #1'); //Correct note test
+        $response->assertSee('Edit'); //edit button present test
+        $response->assertSee('Delete'); //delete button present test
+    }
+
+    public function test_note_viewability_from_unathorized_user()
+    {
+        $this->actingAs($this->userTwo);
```
Comment (fw-aaron) [inline tests/Feature/MyTests/NoteTest.php:111]: Why is this user not authorized?

**Review by fw-aaron — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/NoteController.php (hunk 1 above)
Comment (fw-aaron) [inline app/Http/Controllers/NoteController.php:14]: (This comment doesn't have to do with this line of code. I'm just creating a "conversation" that can be "resolved".)

If I'm not running `npm run dev`, the tests pass. Can you make them pass without running `npm run dev`?

File: tests/Feature/MyTests/NoteTest.php (hunk 2 above)
Comment (DevonScoles) [reply tests/Feature/MyTests/NoteTest.php]: Thank you for pointing that out! I definitely see how this could be referencing other users if I only rely on their id. And if there was a lot of users I might not even know the users id that I'm referencing. In the future I should definitely reference their name or email like you said:
`$userOne = User::where('name', 'Test User #1')->first();`
`$userTwo = User::where('name', 'Test User #2')->first();`

File: tests/Feature/MyTests/NoteTest.php (hunk 3 above)
Comment (DevonScoles) [reply tests/Feature/MyTests/NoteTest.php:92]: Ah I see, I understand how that would make more sense since the homepage of the application is `/note`. It would've been better to test that `/note` redirects if the user isn't logged in rather than testing `/` since it will redirect on `/` whether or not the user is logged in.

File: tests/Feature/MyTests/NoteTest.php (hunk 4 above)
Comment (DevonScoles) [reply tests/Feature/MyTests/NoteTest.php:111]: Here I believe I named the test method/function incorrectly. This test is meant to show that a user cannot access a different user's note, not that any of the users are actually unauthorized. A better name for the method could be
`test_user_cannot_access_other_users_note()`

File: app/Http/Controllers/NoteController.php (hunk 1 above)
Comment (DevonScoles) [reply app/Http/Controllers/NoteController.php:14]: Yes the tests still pass without running npm run dev. Forgive me if I overexplain things you already know. With my application npm run dev is used to run the Vite Development server. This compiles my assets I'm using: Tailwind CSS, alpine.js, and PostCSS. It also allows for Hot Module Replacement (HMR) so that when I make a change to the code it's reflected in the browser immediately.

Comment (fw-aaron) [reply app/Http/Controllers/NoteController.php:14]: When I run `php artisan test` without `npm run dev` running, I get

```
php artisan test                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                  
   PASS  Tests\Unit\ExampleTest                                                                                                                                                                                                                                                   
  ✓ that true is true                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                  
   FAIL  Tests\Feature\Auth\AuthenticationTest                                                                                                                                                                                                                                    
  ⨯ login screen can be rendered                                                                                                                                                                                                                                           0.30s  
  ✓ users can authenticate using the login screen                                                                                                                                                                                                                          0.02s  
  ✓ users can not authenticate with invalid password                                                                                                                                                                                                                       0.21s  
  ✓ users can logout                                                                                                                                                                                                                                                       0.01s  
                                                                                                                                                                                                                                                                                  
   FAIL  Tests\Feature\Auth\EmailVerificationTest                                                                                                                                                                                                                                 
  ⨯ email verification screen can be rendered                                                                                                                                                                                                                              0.19s  
  ✓ email can be verified                                                                                                                                                                                                                                                  0.01s  
  ✓ email is not verified with invalid hash                                                                                                                                                                                                                                0.20s  
                                                                                                                                                                                                                                                                                  
   FAIL  Tests\Feature\Auth\PasswordConfirmationTest                                                                                                                                                                                                                              
  ⨯ confirm password screen can be rendered                                                                                                                                                                                                                                0.19s  
  ✓ password can be confirmed                                                                                                                                                                                                                                              0.01s  
  ✓ password is not confirmed with invalid password                                                                                                                                                                                                                        0.21s  
                                                                                                                                                                                                                                                                                  
   FAIL  Tests\Feature\Auth\PasswordResetTest                                                                                                                                                                                                                                     
  ⨯ reset password link screen can be rendered                                                                                                                                                                                                                             0.19s  
  ✓ reset password link can be requested                                                                                                                                                                                                                                   0.01s  
  ⨯ reset password screen can be rendered                                                                                                                                                                                                                                  0.21s  
  ✓ password can be reset with valid token                                                                                                                                                                                                                                 0.01s  
                                                                                                                                                                                                                                                                                  
   PASS  Tests\Feature\Auth\PasswordUpdateTest                                                                                                                                                                                                                                    
  ✓ password can be updated                                                                                                                                                                                                                                                0.01s  
  ✓ correct password must be provided to update password                                                                                                                                                                                                                   0.01s  
                                                                                                                                                                                                                                                                                  
   FAIL  Tests\Feature\Auth\RegistrationTest                                                                                                                                                                                                                                      
  ⨯ registration screen can be rendered                                                                                                                                                                                                                                    0.19s  
  ✓ new users can register                                                                                                                                                                                                                                                 0.03s  
                                                                                                                                                                                                                                                                                  
   FAIL  Tests\Feature\NoteTest                                                                                                                                                                                                                                                   
  ⨯ application response if user isnt logged in                                                                                                                                                                                                                            0.21s  
  ⨯ application response if user is logged in                                                                                                                                                                                                                              0.21s  
  ⨯ note viewability and validity                                                                                                                                                                                                                                          0.21s  
  ⨯ user cannot access other users note                                                                                                                                                                                                                                    0.22s  
  - new note button → This test just tests the route /note/create which is used by the button                                                                                                                                                                              0.02s  
  ⨯ note creation                                                                                                                                                                                                                                                          0.21s  
  - empty note creation → This test is just an edge case which actually simulates the validate method from NoteController failing and redirecting                                                                                                                          0.04s  
  - oversized note creation → Test purley depends on memory size                                                                                                                                                                                                           0.02s  
  - sql injection prevention → Laravell's Eloquent ORM is used in the models so it automatically protects against SQL injection                                                                                                                                            0.02s  
  ⨯ note update                                                                                                                                                                                                                                                            0.22s  
  - note edit button → Tests the route /note/noteid which is done in the test_note_update_method anyways                                                                                                                                                                   0.02s  
  - application handles special characters → Edge case to make sure notes with <>{}[] can still be created                                                                                                                                                                 0.02s  
  ⨯ note delete method                                                                                                                                                                                                                                                     0.23s  
  - user cannot delete other users note → One user cannot view another\'s notes in order to delete them. CSRF blocks then from using backend methods as well                                                                                                               0.02s  
                                                                                                                                                                                                                                                                                  
   FAIL  Tests\Feature\ProfileTest                                                                                                                                                                                                                                                
  ⨯ profile page is displayed                                                                                                                                                                                                                                              0.20s  
  ✓ profile information can be updated                                                                                                                                                                                                                                     0.01s  
  ✓ email verification status is unchanged when the email address is unchanged                                                                                                                                                                                             0.01s  
  ✓ user can delete their account                                                                                                                                                                                                                                          0.01s  
  ✓ correct password must be provided to delete account                                                                                                                                                                                                                    0.01s  
  ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────  
   FAILED  Tests\Feature\Auth\AuthenticationTest > login screen can be rendered                                                                                                                                                                                                   
  Expected response status code [200] but received 500.                                                                                                                                                                                                                           
Failed asserting that 500 is identical to 200.                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                  
The following exception occurred during the last request:                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                                  
Illuminate\Foundation\ViteManifestNotFoundException: Vite manifest not found at: /app/public/build/manifest.json in /app/vendor/laravel/framework/src/Illuminate/Foundation/Vite.php:727
```

Comment (fw-aaron) [inline app/Http/Controllers/NoteController.php:14]: When I run the application, it prompts me to log in. I create a new account then I'm always forwarded to `/verify-email`. I didn't get an e-mail. How do I get past this to use the application?

**Review by fw-aaron — CHANGES_REQUESTED** [review]
